### DISCUSSION TITLE
[Question]Is caching torch.split views of the encoder output (pinning the whole concatenated buffer) an intentional tradeoff? The per-item-list escape hatch seems to have ~0 encoder coverage

### DISCUSSION THREAD

**Discussion by wly-115** [sgl-project/sglang#39056] [Q&A]
Category: Q&A

---

## What I'm trying to understand

`MultiModalStaticCache` caches each embedding as a `torch.split` **view** into the encoder's single concatenated output buffer. The `DataEmbeddingFunc` docstring already acknowledges the consequence:

> *"a torch.split view would pin the whole concatenated buffer for as long as any single item stays cached"* — `python/sglang/srt/managers/mm_schedule.py` (L135-142)

and recommends the per-item-**list** return form so that *"each cached entry then owns its storage instead of being a view pinning the concatenated buffer."*

Since the comment clearly saw this coming, my question is: **is the current view-based caching a deliberate tradeoff (avoid a copy, lean on encoders returning lists), or an oversight?** What prompts the question is that, as far as I can tell, ~no real encoder actually takes the safe list path — so the recommended escape hatch appears to have near-zero coverage in practice.

## What I looked at (so I'm not guessing)

**Cache side** — `python/sglang/srt/mem_cache/multimodal_cache.py`:
- `set()` (L102-122): no `.clone()` / `.contiguous()`; the `EmbeddingResult(embedding=...)` is inserted as-is.
- `_get_tensor_size()` (L67-68): `return embedding.element_size() * embedding.numel()` — measures the **view's** logical size, not the pinned parent storage.

**Where the views come from** — `python/sglang/srt/managers/mm_schedule.py`:
- `_batch_encode_per_image_misses` (L396-401): `reshape(...)` → `torch.split(all_miss_embedding, token_counts, dim=0)` → `embedding_cache.set(...)`.
- `_get_chunked_embedding_by_item` (L473-482): same `torch.split` → `set(...)`.
- The safe branch only triggers when `data_embedding_func` returns a `list` (`_flatten_embedding_result`, L149-164).

**Encoder return forms** — I statically checked 30 model files under `python/sglang/srt/models/` for the scheduler-facing `get_image_feature` / `get_video_feature` / `get_audio_feature`:

| Return form | Files |
|---|---|
| single concatenated tensor → `torch.split` → view | **24** |
| per-item **list** (safe / owns storage) | **0** |
| no scheduler embedding method in that file | 4 |

Every mainstream VLM I checked returns a single concatenated tensor: **Qwen2-VL, Qwen2.5-VL, Qwen3-VL, InternVL, Kimi-VL, LLaVA, MiniCPM-V, Gemma3, Pixtral, Phi4, GLM-4V**. Even the audio encoders the docstring cites as the motivating per-item example (`qwen2_audio`, `gemma3n_mm`, `minicpmo`, `phi4mm` `get_audio_feature`) all end in `torch.cat(...)`. The only `List[torch.Tensor]` return I found (`KimiK3VisionTower.forward`) is not the scheduler entry point, and its downstream projector re-concatenates.

*Caveat, stated honestly:* I only sampled 30 of ~68 candidate model files, and a few (`qwen3_omni_moe` image/video, `gemma3n_audio`) inherit or live in a parent file I didn't fully trace. This is static reading, not a runtime memory measurement — corrections very welcome.

## The two things I'd like to confirm or be corrected on

1. **Pinning**: with a mostly-cached multi-image request, does one surviving cache entry keep the whole per-batch concatenated encoder buffer resident (real footprint ≫ logical size of the retained items)? Is that acceptable in practice, e.g. because a batch's items are usually evicted together?

2. **LRU accounting**: `_get_tensor_size` charges the view's `numel()`, so `current_size` seems to undercount the pinned parent storage — meaning the cache could exceed the configured `max_size` without noticing. Is this known / intended?

## Relationship to #33388

I believe this shares the *root cause* of #33388 (views pinning parent storage) but on a **different code path**: #33388 is the pickle/serialization side (`get_new_expanded_mm_items` / `_slice_value`), whereas this is the in-process embedding cache (`mm_schedule.py` + `multimodal_cache.py`). A fix to #33388's serialization path would not cover this one. Note there is **no 518x pickle amplification** here — this cache isn't serialized cross-process, so the concern is only pinning + accounting, not multiplication.

## If it turns out worth changing

Two independent, low-risk directions (only if you agree it's not intended):
1. **Materialize on insert** — `.clone()` in `set()` (or right after `torch.split`) so each entry owns compact storage. Costs one copy per *miss* (not per hit), cheap next to re-running the encoder.
2. **Fix accounting only** — charge `_get_tensor_size` by real storage footprint so `max_size` is respected, even if the pin stays. Near-zero cost.

**Comment by flymysql** [comment]
## Short answer

Not intentional, and not an oversight either — it was a **real accounting + retention bug, which has now been fixed on `main`**. Your analysis was correct on both points, and both fixes you proposed are essentially the fix that landed.

## The fix landed 2026-09-13

- **PR:** [#39120 — "Fix multimodal embedding cache retaining full batches through views"](https://github.com/sgl-project/sglang/pull/39120)
- **Commit:** `24b6c1c7f5ba6c95a96acc9e77b4f5efb23eff97`
- **Status:** merged by @cctry on 2026-09-13, ~2 days after you opened this

Your discussion opened 2026-09-11 (08:08 UTC); the PR was created the same day at 18:55 UTC, ~11 hours later, and merged 2026-09-13.

One caveat on that timeline: **PR #39120 does not reference this discussion**, and neither does its commit message, so I can't confirm whether your report prompted it or the two were independent. I'm flagging it rather than claiming credit either way — but if it was independent, that only strengthens your read that this was a known-shaped problem.

## What changed, mapped onto your two questions

### Your question 2 — LRU accounting → fixed

`_get_tensor_size` was measuring the **view**, exactly as you described. It now measures the backing storage:

```python
# before
def _get_tensor_size(embedding: torch.Tensor):
    return embedding.element_size() * embedding.numel()

# after (main)
def _get_tensor_size(embedding: torch.Tensor):
    return embedding.untyped_storage().nbytes()
```

`set()` now evicts against the larger of (real storage, logical payload) and charges the **real** size on insert, so `max_size` can no longer be silently exceeded:

```python
tensor = embedding.embedding
storage_size = _get_tensor_size(tensor)
data_size = min(storage_size, tensor.element_size() * tensor.numel())
while self.current_size + data_size > self.max_size:
    ...
```

### Your question 1 — pinning → fixed by the first option you listed

It materializes on insert, but **conditionally** — only when the storage actually exceeds the payload, so full-storage tensors, reshapes and duplicate keys avoid a pointless copy:

```python
if storage_size > data_size:
    # Clone admitted slices so one cached item cannot retain a whole batch.
    embedding = replace(embedding, embedding=tensor.clone())
data_size = _get_tensor_size(embedding.embedding)
```

So: your proposed **option 1 (materialize on insert)** *and* **option 2 (fix accounting only)** were both adopted; the clone is gated so it isn't paid on the common path. The cost is one copy per **miss**, as you argued.

### The escape hatch question

The per-item-list return form still exists, and the docstring you quoted is unchanged in spirit. The maintainers chose to make the **view path safe** rather than rely on encoders opting into the list path — which is consistent with your finding that ~0 encoders take the list route. Your 24/0/4 static sample was the useful signal here.

## The regression test was inverted to match

This is worth checking against your own repro, because the old test literally asserted the buggy behaviour:

```python
# before: asserted split views share one buffer
def test_tensor_cache_entries_share_storage():
    ...
    assert emb.untyped_storage().nbytes() == total_tokens * HIDDEN * emb.element_size()

# after: asserts each entry owns compact storage
def test_tensor_cache_entries_own_storage():
    ...
    assert emb.untyped_storage().nbytes() == emb.numel() * emb.element_size()
```

`test/registered/chunked_prefill/test_mm_chunked_embedding_unit.py`

## Caveats worth knowing before you consume this

1. **Not in any release yet.** Latest release is `v0.5.19` (published 2026-09-05). I verified the `v0.5.19` tag still has the old `_get_tensor_size` returning `element_size() * numel()` and the old unconditional `data_size = _get_tensor_size(...)` in `set()`. So you need `main` (commit `24b6c1c7f` or later) — upgrading to `v0.5.19` will **not** get you the fix.
2. **PR #39120 did not run GPU tests.** Its own test plan says "GPU performance tests were not run. The added copy is limited to admitted tensors whose backing storage exceeds their payload." Accuracy tests were also not run ("Model accuracy tests were not run"), and the registered scheduler suite was blocked locally by a missing `numpy`. Given `pinning` was a memory-footprint concern, a runtime measurement on your repro (does per-batch footprint now track retained items?) would be a genuinely useful follow-up.
3. **Relationship to #33388 is as you described** — different code path. Note that this fix closes the in-process embedding-cache path (`mm_schedule.py` + `multimodal_cache.py`); it does **not** touch the pickle/serialization amplification path in #33388 (`get_new_expanded_mm_items` / `_slice_value`). Your caveat that a #33388 fix would not cover this one was right, and the converse holds too: this fix does not cover #33388.

## One correction to your framing (minor)

The docstring lines you cite (`mm_schedule.py` L135-142) are the `DataEmbeddingFunc` contract comment, which still stands — it recommends the per-item list form. What changed is that `MultiModalStaticCache.set()` no longer *requires* that contract to hold in order to be correct. So the comment isn't stale; the "escape hatch has ~0 coverage" concern simply stopped being load-bearing.

If you want, I can point you at the exact `main` commit to pin, or help re-run your static sample against the fixed `set()` to confirm the 24-file finding is now moot.

---

*Verified against `main` on 2026-09-16: commit `24b6c1c7f`, PR #39120 diff, `v0.5.19` tag contents, `python/sglang/srt/mem_cache/multimodal_cache.py`, `python/sglang/srt/managers/mm_schedule.py`, `test/registered/chunked_prefill/test_mm_chunked_embedding_unit.py`.*

**Reply by wly-115** [reply]
Ah, a little bummed — I'd actually been hoping to fix this one myself😭. But great that it's already handled; #39120 covers both the accounting and the clone-on-admit.
